events/events_test.go

```diff
@@ -0,0 +1,247 @@
+	testDataPath = "${GOPATH}/src/github.com/lytics/gobyairship/events/testdata"
+)
+
+func init() {
+	if os.Getenv("TEST_EVENTS_PATH") != "" {
+		testDataPath = os.Getenv("TEST_EVENTS_PATH")
+	}
+}
+
+// fakeClient impelemnts the Client interface but is backed by a fixture
+// instead of the real Urban Airship API.
+type fakeClient struct {
+	filter events.Type
+	data   io.ReadCloser
+}
+
+func (c *fakeClient) Post(url string, body interface{}) (*http.Response, error) {
+	// body should be a Request; validate it
+	req, ok := body.(*events.Request)
+	if !ok {
+		return nil, fmt.Errorf("body is not a Request: %T", body)
+	}
+	if len(req.Filters) == 0 && c.filter != "all" {
+		return nil, fmt.Errorf("expected filter=%q but no filter specified", c.filter)
+	}
+	if len(req.Filters) > 1 || len(req.Filters[0].Types) != 1 {
+		return nil, fmt.Errorf("expected filter=%q but received filter=%v", c.filter, req.Filters)
+	}
+	if req.Filters[0].Types[0] != c.filter {
+		return nil, fmt.Errorf("expected filter=%q but received filter=%q", c.filter, req.Filters[0].Types[0])
+	}
+	return &http.Response{StatusCode: 200, Body: c.data}, nil
+}
+
+// filter type test files
+var filterTypes = map[string]events.Type{
+	"all":        "",
+	"close":      events.TypeClose,
+	"first_open": events.TypeFirst,
+	"location":   events.TypeLocation,
+	"open":       events.TypeOpen,
+	"send":       events.TypeSend,
+	"tag_change": events.TypeTagChange,
+	"uninstall":  events.TypeUninstall,
+	"push_body":  events.TypePush,
+}
+
+func TestFilterTypes(t *testing.T) {
+	t.Parallel()
+	for fname, ftype := range filterTypes {
+		t.Log("Testing", fname)
+		fn := fmt.Sprintf("%s/%s.json", os.ExpandEnv(testDataPath), fname)
+		f, err := os.Open(fn)
+		if err != nil {
+			t.Fatalf("Error opening filter file %q: %v", fn, err)
+		}
+		fc := &fakeClient{filter: ftype, data: f}
+
+		offset := uint64(0)
+		resp, err := events.Fetch(fc, events.StartOffset, 0, nil, &events.Filter{Types: []events.Type{ftype}})
+		if err != nil {
+			t.Errorf("Received error fetching %s: %v", fname, err)
+			continue
+		}
+
+		// Check all events
+		i := 0
+		for ev := range resp.Events() {
+			i++
+			if ev.Offset < offset {
+				t.Error("%s - Expected offset to monotonically increase: %d < %d", fname, ev.Offset, offset)
+				continue
+			}
+			offset = ev.Offset
+			if msg := checkEvent(ftype, ev); msg != "" {
+				t.Errorf("%s - %s\n%s", fname, msg, string(ev.Body))
+			}
+		}
+		if resp.Err() != io.EOF {
+			t.Errorf("Received error iterating over events for %s: %v", fname, err)
+		}
+		if i == 0 {
+			t.Errorf("No events processed for %s", fname)
+		}
+	}
+}
+
+func checkEvent(ft events.Type, ev *events.Event) (errmsg string) {
+	if ev.ID == "" {
+		return "Missing ID"
+	}
+	// ft == "" means ev may be of /any/ type
+	if ft != "" && ev.Type != ft {
+		return fmt.Sprintf("Expected type %s but found %s", ft, ev.Type)
+	}
+	if ev.Occurred.IsZero() || ev.Occurred.After(time.Now()) {
```

Comment:
should it be `ev.Occurred.After(time.Now().UTC())` then? Probably doesn't matter.